commit 5c458e71ea9ccbfb25d67256fb8a78e00ba8cfb0 (refs/changes/47/2447/4)
Author: Jordan Hendricks <jordan.hendricks@joyent.com>
Date:   2017-08-23T22:57:59+00:00 (2 years, 2 months ago)
    
    MANTA-3344 MPU commit doesn't store a muskie `etag` on the Moray record
    Reviewed by:
    Approved by:

diff --git a/lib/uploads/common.js b/lib/uploads/common.js
index e22ca84..2bba4de 100644
--- a/lib/uploads/common.js
+++ b/lib/uploads/common.js
@@ -393,6 +393,9 @@ function createTargetObjectRecord(upload, size, md5, cb) {
                     // createMetadata does most of the work for us here, but
                     // a few values need to be overwritten.
                     md.objectId = upload.get(mdKeys.OBJECT_ID);
+                    assert.ok(md.objectId);
+
+                    md.etag = md.objectId;
                     md.contentLength = size;
                     md.contentMD5 = md5;
 
diff --git a/test/mpu/commit.test.js b/test/mpu/commit.test.js
index 51f454d..d968219 100644
--- a/test/mpu/commit.test.js
+++ b/test/mpu/commit.test.js
@@ -8,6 +8,7 @@
  * Copyright (c) 2017, Joyent, Inc.
  */
 
+var MemoryStream = require('stream').PassThrough;
 var uuid = require('node-uuid');
 var path = require('path');
 var vasync = require('vasync');
@@ -164,6 +165,62 @@ test('commit upload: already commited, same set of parts', function (t) {
 });
 
 
+test('commit upload: ensure etag exists on target object', function (t) {
+    var self = this;
+    self.createUpload(self.path, null, function (err) {
+        if (ifErr(t, err, 'created upload')) {
+            t.end();
+            return;
+        }
+
+        self.commitUpload(self.uploadId, [], function (err2) {
+            if (ifErr(t, err2, 'committed upload')) {
+                t.end();
+                return;
+            }
+
+            self.client.info(self.path, function (err3, info1) {
+                if (ifErr(t, err3, 'HEAD target object')) {
+                    t.end();
+                    return;
+                }
+
+                t.ok(info1);
+                t.ok(info1.etag);
+
+                var s = new MemoryStream();
+                var string = 'foobar';
+                var opts = {
+                    size: string.length
+                };
+                setImmediate(s.end.bind(s, string));
+
+                self.client.put(self.path, s, opts, function (err4, res) {
+                    if (ifErr(t, err4, 'overwriting target object')) {
+                        t.end();
+                        return;
+                    }
+
+                    self.client.info(self.path, function (err5, info2) {
+                        if (ifErr(t, err5, 'HEAD overwritten target object')) {
+                            t.end();
+                            return;
+                        }
+
+                        t.ok(info2);
+                        t.ok(info2.etag);
+                        t.ok(info1.etag !== info2.etag);
+                        t.done();
+                    });
+                });
+            });
+        });
+    });
+});
+
+
+
+
 // Commit: invalid upload (not related to the JSON API inputs)
 
 test('commit upload: already aborted', function (t) {
